plugins/design.cpp

```diff
@@ -0,0 +1,271 @@
+  East,
+  West,
+  DragPoint,
+  MouseOver,
+  InShape,
+  ExtraPoint,
+  NumFlags
+};
+
+uint32_t gen_pen_key(bool n, bool s, bool e, bool w, bool is_drag_point,
+                     bool is_mouse_over, bool inshape, bool extra_point) {
+  std::bitset<static_cast<size_t>(PenMask::NumFlags)> ret;
+  ret[PenMask::North] = n;
+  ret[PenMask::South] = s;
+  ret[PenMask::East] = e;
+  ret[PenMask::West] = w;
+  ret[PenMask::DragPoint] = is_drag_point;
+  ret[PenMask::MouseOver] = is_mouse_over;
+  ret[PenMask::InShape] = inshape;
+  ret[PenMask::ExtraPoint] = extra_point;
+
+  return ret.to_ulong();
+}
+
+Screen::Pen make_pen(const std::pair<int, int> &direction, bool is_drag_point,
+                     bool is_mouse_over, bool inshape, bool extra_point) {
+  color_value color = COLOR_GREEN;
+  int ycursor_mod = 0;
+
+  if (!extra_point) {
+    if (is_drag_point) {
+      color = COLOR_CYAN;
+      ycursor_mod += 6;
+      if (is_mouse_over) {
+        color = COLOR_MAGENTA;
+        ycursor_mod += 3;
+      }
+    }
+  } else {
+    ycursor_mod += 15;
+    color = COLOR_LIGHTRED;
+
+    if (is_mouse_over) {
+      color = COLOR_RED;
+      ycursor_mod += 3;
+    }
+  }
+
+  Screen::Pen pen;
+  pen.ch = inshape ? 'X' : 'o';
+  pen.fg = color;
+  int selected_tile_texpos = 0;
+  Screen::findGraphicsTile("CURSORS", direction.first,
+                           direction.second + ycursor_mod,
+                           &selected_tile_texpos);
+  pen.tile = selected_tile_texpos;
+
+  return pen;
+}
+
+Screen::Pen get_pen(int x, int y, ShapeMap &arr, const std::string &type = "") {
+  bool n = false, w = false, e = false, s = false;
+  if (has_point(x, y)) {
+    if (y == 0 || !has_point(x, y - 1)) n = true;
+    if (x == 0 || !has_point(x - 1, y)) w = true;
+    if (!has_point(x + 1, y)) e = true;  // TODO check map size
+    if (!has_point(x, y + 1)) s = true;  // TODO check map size
+  }
+
+  bool is_drag_point = type == "drag_point";
+  bool is_extra = type == "extra_point";
+  bool is_in_shape = has_point(x, y);
+  auto mouse_pos = Gui::getMousePos();
+  bool mouse_over = mouse_pos.x == x && mouse_pos.y == y;
+
+  uint32_t pen_key =
+      gen_pen_key(n, s, e, w, is_drag_point, mouse_over, is_in_shape, is_extra);
+
+  if (CURSORS_MAP.count({n, w, e, s}) > 0 && has_point(x,y)) {
+    arr[x][y].cursor_coords = CURSORS_MAP.at({n, w, e, s});
+  }
+
+  if (PENS.find(pen_key) == PENS.end()) {
+    std::pair<int, int> cursor{-1, -1};
+
+    if (type != "") {
+      return make_pen(CURSORS_MAP.at({n, w, e, s}), is_drag_point, mouse_over,
+                      is_in_shape, is_extra);
+    }
+
+    if (CURSORS_MAP.count({n, w, e, s}) > 0) {
+      PENS.emplace(pen_key,
+                   make_pen(CURSORS_MAP.at({n, w, e, s}), is_drag_point,
+                            mouse_over, is_in_shape, is_extra));
+      if (type == "" && has_point(x,y)) {
+        arr[x][y].penKey = pen_key;
+      }
+    }
+  }
+
+  // DEBUG(status).print("not cached lmao\n");
```

Comment:
feel free to leave these uncommented. they have very little runtime overhead, and they're useful for debugging when you need them. If a player has an issue, you can ask them to turn on debug logging and give you a copy of the output.

Reply:
Good to know, but this one wouldn't actually be helpful so I'm removing it.